### PR TITLE
Add gettxoutsetinfo hash_type option

### PR DESCRIPTION
**Pull request by fjahr**
This is another intermediate part of the Coinstats Index (tracked in #18000).

Sjors suggested [here](https://github.com/bitcoin/bitcoin/pull/18000#issuecomment-641423019) that the part of the changes in #19145 that don't rely on the new `hash_type` muhash, i.e. that are for `hash_type=none`, could be merged separately from everything involving muhash. So these changes are extracted from #19145 here and can be merged without any other requirements.

Building the index with no UTXO set hash is still valuable because `gettxoutsetinfo` can still be used to audit the `total_amount` for example. By itself this PR is not a huge improvement, `hash_type=none` is speeding up `gettxoutsetinfo` by about 10%, but it enables the implementation of an index on top of it in a follow-up and that means large parts of the index code of Coinstats Index can be merged while reviews for the hashing algorithm might take longer.

### CODE REVIEW

**Comment by fjahr** [comment]
cc @Sjors

File: src/node/coinstats.cpp [hunk 1]
```diff
@@ -42,19 +42,21 @@ static void ApplyStats(CCoinsStats &stats, CHashWriter& ss, const uint256& hash,
 }
 
 //! Calculate statistics about the unspent transaction output set
-bool GetUTXOStats(CCoinsView* view, CCoinsStats& stats, const std::function<void()>& interruption_point)
+template <typename T>
+bool GetUTXOStats(CCoinsView* view, CCoinsStats& stats, T hash_obj, const std::function<void()>& interruption_point)
```
Comment (MarcoFalke) [inline src/node/coinstats.cpp:57]: nit in commit 87dd07e0382c6398d118d88c88b0c9ef24251e37

```suggestion
static bool GetUTXOStats(CCoinsView* view, CCoinsStats& stats, T hash_obj, const std::function<void()>& interruption_point)
```

File: src/node/coinstats.cpp [hunk 2]
```diff
@@ -75,9 +77,33 @@ bool GetUTXOStats(CCoinsView* view, CCoinsStats& stats, const std::function<void
...
+
+bool GetUTXOStats(CCoinsView *view, CCoinsStats &stats, CoinStatsHashType hash_type, const std::function<void()>& interruption_point)
+{
+    switch(hash_type) {
+        case(CoinStatsHashType::HASH_SERIALIZED): {
+            CHashWriter ss(SER_GETHASH, PROTOCOL_VERSION);
+            return GetUTXOStats(view, stats, ss, interruption_point);
+        }
+    }
+    return false; // Unreachable code
```
Comment (MarcoFalke) [inline src/node/coinstats.cpp:111]: 87dd07e0382c6398d118d88c88b0c9ef24251e37:

According to the developer notes, assert(false) and with the appropriate comment

File: src/rpc/client.cpp [hunk 3]
```diff
@@ -123,6 +123,7 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "gettxout", 1, "n" },
     { "gettxout", 2, "include_mempool" },
     { "gettxoutproof", 0, "txids" },
+    { "gettxoutsetinfo", 0, "hash_type" },
```
Comment (MarcoFalke) [inline src/rpc/client.cpp:126]: 87dd07e0382c6398d118d88c88b0c9ef24251e37

why is this needed? Looks like a str to me

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -2316,7 +2331,7 @@ UniValue dumptxoutset(const JSONRPCRequest& request)
 
         ::ChainstateActive().ForceFlushStateToDisk();
 
-        if (!GetUTXOStats(&::ChainstateActive().CoinsDB(), stats, RpcInterruptionPoint)) {
+        if (!GetUTXOStats(&::ChainstateActive().CoinsDB(), stats, CoinStatsHashType::HASH_SERIALIZED, RpcInterruptionPoint)) {
```
Comment (MarcoFalke) [inline src/rpc/blockchain.cpp:2336]: in commit ea41a5bdcc49566be0c23ef9362781e70f557bad:

why is this not changed to NONE?

File: src/node/coinstats.cpp [hunk 5]
```diff
@@ -41,6 +41,17 @@ static void ApplyStats(CCoinsStats &stats, CHashWriter& ss, const uint256& hash,
     ss << VARINT(0u);
 }
 
+static void ApplyStats(CCoinsStats &stats, nullptr_t, const uint256& hash, const std::map<uint32_t, Coin>& outputs)
```
Comment (MarcoFalke) [inline src/node/coinstats.cpp:44]: nit ea41a5bdcc49566be0c23ef9362781e70f557bad

```suggestion
static void ApplyStats(CCoinsStats& stats, nullptr_t, const uint256& hash, const std::map<uint32_t, Coin>& outputs)
```

**Review by maflcko — COMMENTED** [review]
review ACK e884c98996b135eee51e5d3994ae604110c797fd 🚩

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK e884c98996b135eee51e5d3994ae604110c797fd 🚩
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhpPgv/XaXojJucDooSWoRbsa/hlk3xG9yYOVK6BisILFXd1xNoo2pBnGwuSX2L
lq2h19aOP4Zk/iYOzPwdP8UAOaDBoFJ9pAQwo0YYuHu2lIFLlmjFScJMd1Y3OUhl
+d6SAYSsc6lQ9QxpaMwocZ4V6X3JrFcXBK0vCscF2wm/bBVCOwBo/x0sSrPFyO9M
dAnaJIbth+P6NVu/jNKtigPK0hTal2rTqfxB1/RZbnoJw3yugWLPDsF+/FiBp5+Y
g2Y/uSTBp25RS/vKN6+L0qJfuiJGS/ixwUMpD/kDiY7r3hqIlpn8ZHdyxCRRYZgp
OfOC9ZicTgO2EdNrOepxrIwWbnrtWXyPo/AJlMqadk/8MtqNzQ+tivMfapu8gPUi
WzI7W5nYv8ZFTX85rz8nwhDtaxKliIgkkui8rEhpgMESpPYWx5X6443C9ZmRYV0X
XDEXWaLnBNaX5a8dE+uTpzlODjiY8PIgBTpoAk+7qfzO8YpfGKiOZP7++XISEEwf
ZBDVe5go
=SITy
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `23084cd1f58a12c1391d066fc4f3cc18831bc7705d21198fe3a30252ff1f9f0c  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010823084cd1f58a12c1391d066fc4f3cc18831bc7705d21198fe3a30252ff1f9f0cf010043c36ea22e40fec10a904bab213dde008fff0105ae22a8800e7624efd1d7a1ca6202e7c08f120b398caf3b07d137edb1cc1128e5f268529987efc82e1ff4e5d58adadfaf537ff08f1045eeced21f0087d8215caf55c90a50083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0106876ab515914eb5ded02fd4599ab671008f1045eeced20f008a26bd37ede1079bd0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0109cc6f2920871923d69c4d4d53e5da34108f1045eeced21f008ef6b0e3c97becd280083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f99374de05560003a2e4a285413037a708f1045eeced21f0084f0ab72ece0c9adf0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

File: src/rpc/client.cpp (hunk 3 above)
Comment (Sjors) [reply src/rpc/client.cpp:126]: Indeed, this should be dropped (try: `bitcoin-cli gettxoutsetinfo none`, it works without this line).

File: src/rpc/blockchain.cpp (hunk 4 above)
Comment (Sjors) [reply src/rpc/blockchain.cpp:2336]: Agreed, this hash is unused in UTXO snapshots, as judged by the code as well as the hardcoded UTXO set has in `rpc_dumptxoutset.py`.

File: src/rpc/blockchain.cpp [hunk 6]
```diff
@@ -981,7 +983,7 @@ static UniValue gettxoutsetinfo(const JSONRPCRequest& request)
                         {RPCResult::Type::NUM, "transactions", "The number of transactions with unspent outputs"},
                         {RPCResult::Type::NUM, "txouts", "The number of unspent transaction outputs"},
                         {RPCResult::Type::NUM, "bogosize", "A meaningless metric for UTXO set size"},
-                        {RPCResult::Type::STR_HEX, "hash_serialized_2", "The serialized hash"},
+                        {RPCResult::Type::STR_HEX, "hash_serialized_2", "The serialized hash (only present if 'hash_serialized_2' hash_type is chosen)"},
```
Comment (Sjors) [inline src/rpc/blockchain.cpp:986]: Maybe open a separate PR to see if there's support for deprecating `hash_serialized_2` and switch the default to `NONE` (pre-ACK, unless we find someone who actually uses this)

**Review by Sjors — CHANGES_REQUESTED** [review]
Tested e884c98996b135eee51e5d3994ae604110c797fd on macOS 10.15.5

File: src/node/coinstats.cpp (hunk 1 above)
Comment (fjahr) [reply src/node/coinstats.cpp]: done

File: src/node/coinstats.cpp (hunk 2 above)
Comment (fjahr) [reply src/node/coinstats.cpp]: done

File: src/rpc/client.cpp (hunk 3 above)
Comment (fjahr) [reply src/rpc/client.cpp]: done

File: src/rpc/blockchain.cpp (hunk 4 above)
Comment (fjahr) [reply src/rpc/blockchain.cpp]: done

File: src/node/coinstats.cpp (hunk 5 above)
Comment (fjahr) [reply src/node/coinstats.cpp:44]: done

File: src/node/coinstats.cpp [hunk 7]
```diff
@@ -66,9 +88,38 @@ bool GetUTXOStats(CCoinsView* view, CCoinsStats& stats, const std::function<void
...
+    switch(hash_type) {
+        case(CoinStatsHashType::HASH_SERIALIZED): {
+            CHashWriter ss(SER_GETHASH, PROTOCOL_VERSION);
+            return GetUTXOStats(view, stats, ss, interruption_point);
+        }
+        case(CoinStatsHashType::NONE): {
+            return GetUTXOStats(view, stats, nullptr, interruption_point);
+        }
+    }
+    assert(false); // Unreachable code
```
Comment (MarcoFalke) [inline src/node/coinstats.cpp:111]: Should be identical to the other cases to make it grep-able. See dev-notes example:

```cpp
enum class Tabs {
    INFO,
    CONSOLE,
    GRAPH,
    PEERS
};

int GetInt(Tabs tab)
{
    switch (tab) {
    case Tabs::INFO: return 0;
    case Tabs::CONSOLE: return 1;
    case Tabs::GRAPH: return 2;
    case Tabs::PEERS: return 3;
    } // no default case, so the compiler can warn about missing cases
    assert(false);
}
```
```suggestion
    assert(false);
```

File: src/rpc/blockchain.cpp (hunk 6 above)
Comment (fjahr) [reply src/rpc/blockchain.cpp:986]: Hm, at least BTCPayServer uses it for FastSync: https://github.com/btcpayserver/btcpayserver-docker/tree/master/contrib/FastSync. They say in the docs to compare the whole `gettxoutsetinfo` but the hash is really what ensures there is nothing malicious going on. But I am not sure how much that feature is actually used. I will reach out to them to get an opinion.

Comment (MarcoFalke) [reply src/rpc/blockchain.cpp:986]: Is there any benefit of aggressively deprecating this? With the option to select no hash there should be no downside to just leaving the default as is

File: src/node/coinstats.cpp (hunk 7 above)
Comment (fjahr) [reply src/node/coinstats.cpp]: Ok, thanks. Fixed.

File: src/node/coinstats.cpp [hunk 8]
```diff
@@ -75,9 +77,33 @@ bool GetUTXOStats(CCoinsView* view, CCoinsStats& stats, const std::function<void
...
     }
-    stats.hashSerialized = ss.GetHash();
+
+    FinalizeHash(hash_obj, stats);
+
     stats.nDiskSize = view->EstimateSize();
     return true;
 }
+
+bool GetUTXOStats(CCoinsView *view, CCoinsStats& stats, CoinStatsHashType hash_type, const std::function<void()>& interruption_point)
```
Comment (promag) [inline src/node/coinstats.cpp:100]: ed4c7302298385b797000740960b06ff91082ce8

nit `* view`.

File: src/node/coinstats.cpp [hunk 9]
```diff
@@ -75,9 +77,33 @@ bool GetUTXOStats(CCoinsView* view, CCoinsStats& stats, const std::function<void
...
+{
+    switch(hash_type) {
+        case(CoinStatsHashType::HASH_SERIALIZED): {
```
Comment (promag) [inline src/node/coinstats.cpp:103]: ed4c7302298385b797000740960b06ff91082ce8

nit, this format is not used. Also add space after `switch` above.

**Comment by MarcoFalke** [comment]
```
test/fuzz/coins_view.cpp:281:23: error: no matching function for call to 'GetUTXOStats'

                (void)GetUTXOStats(&coins_view_cache, stats);

                      ^~~~~~~~~~~~

./node/coinstats.h:39:6: note: candidate function not viable: requires at least 3 arguments, but 2 were provided

bool GetUTXOStats(CCoinsView* view, CCoinsStats& stats, const CoinStatsHashType hash_type, const std::function<void()>& interruption_point = {});

     ^

1 error generated.

File: src/node/coinstats.cpp (hunk 9 above)
Comment (MarcoFalke) [reply src/node/coinstats.cpp:103]: ;)
https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy

Comment (fjahr) [reply src/node/coinstats.cpp]: done

**Comment by fjahr** [comment]
Fixed review comments plus some more formatting improvements

File: src/rpc/blockchain.cpp [hunk 10]
```diff
@@ -996,14 +998,29 @@ static UniValue gettxoutsetinfo(const JSONRPCRequest& request)
     CCoinsStats stats;
     ::ChainstateActive().ForceFlushStateToDisk();
 
+    CoinStatsHashType hash_type = CoinStatsHashType::HASH_SERIALIZED;
```
Comment (promag) [inline src/rpc/blockchain.cpp:1001]: nit, move code below to new function and here make it `ParseHashType(request.params[0], CoinStatsHashType::HASH_SERIALIZED)` (2nd is default value if 1st is null).

File: src/rpc/blockchain.cpp [hunk 11]
```diff
@@ -996,14 +998,29 @@ static UniValue gettxoutsetinfo(const JSONRPCRequest& request)
...
+    if (!request.params[0].isNull()) {
+        std::string hash_type_input = request.params[0].get_str();
+
+        if (hash_type_input == "hash_serialized_2") {
+            // keep default
```
Comment (promag) [inline src/rpc/blockchain.cpp:1006]: nit, I'd explicitly set `hash_type = CoinStatsHashType::HASH_SERIALIZED`.

File: src/node/coinstats.cpp [hunk 12]
```diff
@@ -26,26 +36,38 @@ static void ApplyStats(CCoinsStats &stats, CHashWriter& ss, const uint256& hash,
...
         stats.nTransactionOutputs++;
         stats.nTotalAmount += output.second.out.nValue;
-        stats.nBogoSize += 32 /* txid */ + 4 /* vout index */ + 4 /* height + coinbase */ + 8 /* amount */ +
-                           2 /* scriptPubKey len */ + output.second.out.scriptPubKey.size() /* scriptPubKey */;
+        stats.nBogoSize += GetBogoSize(output.second.out.scriptPubKey);
     }
     ss << VARINT(0u);
 }
 
+static void ApplyStats(CCoinsStats& stats, nullptr_t, const uint256& hash, const std::map<uint32_t, Coin>& outputs)
```
Comment (MarcoFalke) [inline src/node/coinstats.cpp:44]:
```
node/coinstats.cpp:44:44: error: unknown type name 'nullptr_t'; did you mean 'std::nullptr_t'?

static void ApplyStats(CCoinsStats& stats, nullptr_t, const uint256& hash, const std::map<uint32_t, Coin>& outputs)

                                           ^~~~~~~~~

                                           std::nullptr_t

/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/x86_64-linux-gnu/c++/9/bits/c++config.h:258:29: note: 'std::nullptr_t' declared here

  typedef decltype(nullptr)     nullptr_t;

                                ^

node/coinstats.cpp:119:25: error: use of undeclared identifier 'nullptr_t'; did you mean 'nullptr'?

static void PrepareHash(nullptr_t, CCoinsStats& stats) {}

                        ^~~~~~~~~

                        nullptr

File: src/rpc/blockchain.cpp (hunk 11 above)
Comment (fjahr) [reply src/rpc/blockchain.cpp]: done

File: src/rpc/blockchain.cpp (hunk 10 above)
Comment (fjahr) [reply src/rpc/blockchain.cpp]: done

**Comment by fjahr** [comment]
Addressed comments and CI failure.

File: src/node/coinstats.cpp (hunk 12 above)
Comment (fjahr) [reply src/node/coinstats.cpp:44]: fixed

File: src/rpc/blockchain.cpp (hunk 6 above)
Comment (Sjors) [reply src/rpc/blockchain.cpp:986]: Most people won't realise how slow this command is by default until they already issued it.

File: src/node/coinstats.cpp [hunk 13]
```diff
@@ -41,6 +41,17 @@ static void ApplyStats(CCoinsStats& stats, CHashWriter& ss, const uint256& hash,
     ss << VARINT(0u);
 }
 
+static void ApplyStats(CCoinsStats& stats, std::nullptr_t, const uint256& hash, const std::map<uint32_t, Coin>& outputs)
+{
+    assert(!outputs.empty());
+    stats.nTransactions++;
```
Comment (Sjors) [inline src/node/coinstats.cpp:47]: I'd like to avoid this duplication, but it can wait.

**Comment by Sjors** [comment]
tACK 40506bf93f955adfbc446c4d5fee4fa8bcfd7d9a

**Review by Sjors — APPROVED** [review]
tACK 40506bf93f955adfbc446c4d5fee4fa8bcfd7d9a

**Comment by MarcoFalke** [comment]
Checked that 40506bf93 is faster with the `none` option on my arm machine. Haven't yet reviewed the code of that commit

File: src/node/coinstats.cpp (hunk 13 above)
Comment (fjahr) [reply src/node/coinstats.cpp:47]: I now remove it in the follow-up when adding hash_type MUHASH https://github.com/bitcoin/bitcoin/pull/19145/files#diff-0ab7b981bd35c1a1f45b5ed75013a698

**Comment by MarcoFalke** [comment]
ACK 40506bf93f955adfbc446c4d5fee4fa8bcfd7d9a 🖨

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 40506bf93f955adfbc446c4d5fee4fa8bcfd7d9a 🖨
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiC+Qv/UfSTrJYS3nGpNjOzD1xGJWB/hTnUlgwgMIm4UkrnJw5tpLgSxniPsa7J
8sK5CaKfpLKpvXjW+sFMcadgHX8IeZlFfvlDev8Bc6T0RrGh/61qgjwKqQUKtAh+
zv+v+QnvkXInpgTYK3oLF7UvtBqEqwpRi04e6m23eLaDcdMC03iBAjYPd+/4WbFV
+wQdWyZTn52Fe3IKR6+QcyQXXZS4lgUYd1K0N90w1nfB1kVx7C04KMn3Ubj6FxOb
kZtIdYLciOCPodgy5idcZGHHEcEg88z3lhXp5JqQmh96dhDf9ehsclgl5ycFVl8I
Iu8BXhHs/OwYg+sy/9IDvdN6uwzknwIv1E9ueSe5fS0Xh8W6KLOaEIa7iAz4OPlQ
fvseGpqHfuuGv8QHK+3e4+GorghSLM2hOyUure5y+l64dWa5BQeubc2FXHFupyf8
pAYWjmUjSnk240Fb4nqRqp3j6iu7joHWFKK5FSOCofZ6pLlZ7cIDluKXsYYqaDJq
p1ypj/n5
=eqXn
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `4ce439fee8be758f1a7c060666d9c236d07dd151ec5a09fed5df26d9df431ae9  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401084ce439fee8be758f1a7c060666d9c236d07dd151ec5a09fed5df26d9df431ae9f010efd34ace6ed456231acd43e3e456700c08fff010074997c5706c82ebe7dc66885e21a16508f1045f0313b0f0080bfcf221f26428630083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01082a2cd324a4efcfed3d1d5f46a28061a08f1045f0313aff008de1c515d5234c18c0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0108c43f0984fab544776d809b00268f9e908f1045f0313aff008ce299776b2f37cc30083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f01095fb4d7094a0e0b25515cc29c91d86a308f1045f0313b0f0080c0763802f98431a0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

File: src/rpc/util.cpp [hunk 14]
```diff
@@ -113,6 +113,23 @@ std::vector<unsigned char> ParseHexO(const UniValue& o, std::string strKey)
     return ParseHexV(find_value(o, strKey), strKey);
 }
 
+CoinStatsHashType ParseHashType(const UniValue& param, const CoinStatsHashType default_type)
+{
+    if (param.isNull()) {
+        return default_type;
+    } else {
```
Comment (MarcoFalke) [inline src/rpc/util.cpp:120]: style-nit: No need for the `else` and all the whitespace padding below

```suggestion
    }
    const std::str...
```

Comment (fjahr) [reply src/rpc/util.cpp:120]: Thanks, now fixed in #19145.
